### PR TITLE
ARCO-150: submit existing tx with new callback

### PR DESCRIPTION
**Pull request by pawellewandowski98**
# Description
- Add new callback to tx when submitting already existing one

### CODE REVIEW

File: internal/metamorph/processor.go [hunk 1]
```diff
@@ -719,3 +720,22 @@ func (p *Processor) storeData(ctx context.Context, data *store.StoreData) error
 	data.LastSubmittedAt = p.now()
 	return p.store.Set(ctx, data)
 }
+
+func addNewCallback(data, reqData *store.StoreData) {
+	if reqData.Callbacks == nil {
+		return
+	}
+	reqCallback := reqData.Callbacks[0]
+	if reqCallback.CallbackURL != "" && reqCallback.CallbackToken != "" && !callbackExists(reqCallback, data) {
```
Comment (boecklim) [inline internal/metamorph/processor.go:729]: We should add the callback also if there is no callback token in the request I think.
That would mean the condition should be this:
```
if reqCallback.CallbackURL != "" && !callbackExists(reqCallback, data)
```

File: internal/metamorph/store/postgresql/postgres.go
Comment (boecklim) [inline internal/metamorph/store/postgresql/postgres.go:1]: Would be nice to have an integration test for this here: `internal/metamorph/store/postgresql/postgres_test.go`

File: internal/metamorph/processor.go [hunk 2]
```diff
@@ -719,3 +720,22 @@ func (p *Processor) storeData(ctx context.Context, data *store.StoreData) error
...
+		data.Callbacks = append(data.Callbacks, reqCallback)
+	}
+}
+
+func callbackExists(callback store.StoreCallback, data *store.StoreData) bool {
+	for _, c := range data.Callbacks {
+		if c.CallbackURL == callback.CallbackURL && c.CallbackToken == callback.CallbackToken {
```
Comment (arkadiuszos4chain) [inline internal/metamorph/processor.go:736]:
```suggestion
		if c == callback {
```

File: test/endpoint_test.go [hunk 3]
```diff
@@ -174,10 +174,22 @@ func createTxChain(privateKey string, utxo0 NodeUnspentUtxo, length int) ([]*bt.
 
 func TestPostCallbackToken(t *testing.T) {
 	tt := []struct {
-		name string
+		name                       string
+		numberOfCallbacks          int
```
Comment (arkadiuszos4chain) [inline test/endpoint_test.go:178]:
```suggestion
		numberOfCallbackServers          int
```

File: internal/metamorph/processor.go (hunk 1 above)
Comment (pawellewandowski98) [reply internal/metamorph/processor.go]: Done

File: internal/metamorph/store/postgresql/postgres.go
Comment (pawellewandowski98) [reply internal/metamorph/store/postgresql/postgres.go:1]: Done

File: internal/metamorph/processor.go (hunk 2 above)
Comment (pawellewandowski98) [reply internal/metamorph/processor.go]: Done

File: test/endpoint_test.go (hunk 3 above)
Comment (pawellewandowski98) [reply test/endpoint_test.go]: Done

**Review by arkadiuszos4chain — APPROVED** [review]

**Review by kuba-4chain — APPROVED** [review]

**Review by boecklim — APPROVED** [review]

**Pull request merged**